ADVANCED_EXAMPLES.md

```diff
@@ -0,0 +1,409 @@
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [`object.getStamp()`. 2 ways.](#objectgetstamp-2-ways)
+  - [First way](#first-way)
+  - [Second way](#second-way)
+- [Self cloneable objects. 3 ways.](#self-cloneable-objects-3-ways)
+  - [A composable stamp which adds `.clone()` to objects](#a-composable-stamp-which-adds-clone-to-objects)
+  - [Another way of self cloning](#another-way-of-self-cloning)
+  - [Memory efficient cloning](#memory-efficient-cloning)
+- [Delayed object instantiation using Promises](#delayed-object-instantiation-using-promises)
+- [Dependency injection tips](#dependency-injection-tips)
+- [Validate before a function call](#validate-before-a-function-call)
+- [EventEmitter without inheritance (`convertConstructor`)](#eventemitter-without-inheritance-convertconstructor)
+- [Hacking stamps](#hacking-stamps)
+  - ["Default" properties](#default-properties)
+  - ["Default" properties as composable behavior](#default-properties-as-composable-behavior)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+
+## `object.getStamp()`. 2 ways.
+
+> Run the examples below for yourself:
+```sh
+$ git clone https://github.com/stampit-org/stampit.git
+$ node stampit/advanced-examples/self-aware.js
+```
+
+You can add `.getStamp()` function to each object with ease. 
+
+First, let's assume you have a following stamp:
+```js
+var User = stampit();
+```
+
+### First way
+Just compose the following stamp to any other stamp.
+```js
+var SelfAware1 = stampit.init(function (ctx) { // the context, has `stamp` property
+  this.getStamp = function () { return ctx.stamp; }
+});
+```
+Let's compose it with the `User` stamp from above:
+```js
+var SelfAwareUser1 = User.compose(SelfAware1);
+```
+Now, let's create a user and call the `.getStamp()`:
+```js
+var user1 = SelfAwareUser1();
+assert.strictEqual(user1.getStamp(), SelfAwareUser1); // All good
+```
+So, now every object instance returns the exact stamp it was built with. Nice!
+
+### Second way
+Another composable stamp which does the same but in a memory efficient way.
+It attaches the function to the `.prototype` of the objects, but not to each one.
+```js
+var SelfAware2 = stampit.init(function (ctx) { // the context, has `stamp` property
+  if (!ctx.stamp.fixed.methods.getStamp) { // Let's add the method only once.
+    var stamp = ctx.stamp; // We should reference only one object, otherwise it is memory leak risky
```

Comment:
"memory leak risky" is not good grammar. Can you explain what you mean by memory leak risky? also